taichi/backends/dx/dx_device.cpp

```diff
@@ -0,0 +1,464 @@
+  TI_NOT_IMPLEMENTED;
+}
+
+void DxResourceBinder::rw_buffer(uint32_t set,
+                                 uint32_t binding,
+                                 DeviceAllocation alloc) {
+  TI_ASSERT_INFO(set == 0, "DX only supports set = 0, requested set = {}", set);
+  ID3D11UnorderedAccessView *uav = g_binding2uav[alloc.alloc_id];
+  binding_map_[binding] = uav;
+}
+
+void DxResourceBinder::buffer(uint32_t set,
+                              uint32_t binding,
+                              DevicePtr ptr,
+                              size_t size) {
+  TI_NOT_IMPLEMENTED;
+}
+
+void DxResourceBinder::buffer(uint32_t set,
+                              uint32_t binding,
+                              DeviceAllocation alloc) {
+  rw_buffer(set, binding, alloc);
+}
+  
+void DxResourceBinder::image(uint32_t set,
+                             uint32_t binding,
+                             DeviceAllocation alloc,
+                             ImageSamplerConfig sampler_config) {
+  TI_NOT_IMPLEMENTED;
+}
+
+// Set vertex buffer (not implemented in compute only device)
+void DxResourceBinder::vertex_buffer(DevicePtr ptr, uint32_t binding) {
+  TI_NOT_IMPLEMENTED;
+}
+
+// Set index buffer (not implemented in compute only device)
+// index_width = 4 -> uint32 index
+// index_width = 2 -> uint16 index
+void DxResourceBinder::index_buffer(DevicePtr ptr, size_t index_width) {
+  TI_NOT_IMPLEMENTED;
+}
+
+DxPipeline::DxPipeline(const PipelineSourceDesc &desc, const std::string &name)
+    : compute_shader_(nullptr) {
+  TI_ASSERT(desc.type == PipelineSourceType::hlsl_src ||
+            desc.type == PipelineSourceType::spirv_binary);
+
+  if (desc.type == PipelineSourceType::hlsl_src) {
+    TI_NOT_IMPLEMENTED;
+  } else {
+    std::vector<uint32_t> spirv_binary(
+        (uint32_t *)desc.data, (uint32_t *)((uint8_t *)desc.data + desc.size));
+    spirv_cross::CompilerHLSL hlsl(std::move(spirv_binary));
+    spirv_cross::CompilerHLSL::Options options;
+    options.shader_model = 40;
+    hlsl.set_hlsl_options(options);
+
+    std::string source = hlsl.compile();
+    TI_TRACE("hlsl source: \n{}", source);
+
+    ID3DBlob *shader_blob;
+    ID3D11Device *device = GetD3D11Device();
+    HRESULT hr = CompileComputeShaderFromString(source, "main",
+                                                device, &shader_blob);
+    if (SUCCEEDED(hr)) {
+      hr = device->CreateComputeShader(shader_blob->GetBufferPointer(),
+                                       shader_blob->GetBufferSize(), nullptr,
+                                       &compute_shader_);
+      shader_blob->Release();
+      compute_shader_->SetPrivateData(WKPDID_D3DDebugObjectName,
+                                      name.size(), name.c_str());
+    } else {
+      TI_ERROR("HLSL compute shader compilation error");
+    }
+  }
+}
+
+DxPipeline::~DxPipeline() {
+}
+
+ResourceBinder *DxPipeline::resource_binder() {
+  return &binder_;
+}
+
+DxCommandList::~DxCommandList() {
```

Comment:
https://docs.microsoft.com/en-us/windows/win32/direct3d11/overviews-direct3d-11-render-multi-thread-command-list In the future we should use DX11's built-in command list instead of emulating it on the CPU. This is a possible improvement, don't need to do it rn. Let's get the back end merged first

Reply:
Acknowledged.